### PR TITLE
Fixed convergence issues with CPU offloading

### PR DESCRIPTION
**Pull request by sanandaraj5597**
Fixed convergence issues when CPU offloading is enabled.

1) When activation offloading was enabled, the reloaded weights during backprop was re-constructed into a torch.nn.Parameter with the `requires_grad` switch as false. This switch should be try to compute weight gradients. Fixed this.

2) When weights are offloaded we try to offload the gradient accumulation buffer, by doing a deep copy of this buffer and recreating this which confuses the Distributed Optimizer about this buffer location. This makes the DistOpt using a stale location to look for fresh gradients from recent micro-batches. This has been fixed by not offloading the gradient accumulation buffer. Offloading gradient accumulation buffer was considered a free lunch but this subtle issues made me go back with my decision.

### CODE REVIEW

File: transformer_engine/pytorch/module/layernorm_linear.py [hunk 1]
```diff
@@ -411,7 +409,7 @@ def backward(
             )
 
             if ctx.cpu_offloading and ctx.fuse_wgrad_accumulation:
-                weight = torch.nn.Parameter(weight, False)
+                weight = torch.nn.Parameter(weight)
```
Comment (ksivaman) [inline transformer_engine/pytorch/module/layernorm_linear.py:412]:
```suggestion
                weight = torch.nn.Parameter(weight.requires_grad)
```

Comment (ksivaman) [reply transformer_engine/pytorch/module/layernorm_linear.py:412]: Isn't it best in this case to simply retain the value.

File: transformer_engine/pytorch/module/layernorm_mlp.py [hunk 2]
```diff
@@ -570,8 +567,8 @@ def backward(
             )
 
             if ctx.cpu_offloading and ctx.fuse_wgrad_accumulation:
-                fc1_weight = Parameter(fc1_weight, False)
-                fc2_weight = Parameter(fc2_weight, False)
+                fc1_weight = Parameter(fc1_weight)
+                fc2_weight = Parameter(fc2_weight)
```
Comment (ksivaman) [inline transformer_engine/pytorch/module/layernorm_mlp.py:571]:
```suggestion
                fc1_weight = Parameter(fc1_weight.requires_grad)
                fc2_weight = Parameter(fc2_weight.requires_grad)
```

File: transformer_engine/pytorch/module/linear.py [hunk 3]
```diff
@@ -403,7 +401,7 @@ def backward(ctx, grad_output: torch.Tensor) -> Tuple[Union[torch.Tensor, None],
             )
 
             if ctx.cpu_offloading and ctx.fuse_wgrad_accumulation:
-                weight = torch.nn.Parameter(weight, False)
+                weight = torch.nn.Parameter(weight)
```
Comment (ksivaman) [inline transformer_engine/pytorch/module/linear.py:404]:
```suggestion
                weight = torch.nn.Parameter(weight.requires_grad)
```

File: transformer_engine/pytorch/module/layernorm_linear.py (hunk 1 above)
Comment (sanandaraj5597) [reply transformer_engine/pytorch/module/layernorm_linear.py:412]: Hmm, CPU offloading works only for training. So, it is expected to have requires_grad=True. So, wanted to make it explicit and clear.

Comment (ksivaman) [reply transformer_engine/pytorch/module/layernorm_linear.py:412]: If CPU offloading is enabled, then we should check for `weight.requires_grad` being true before instead of explicitly setting it here even if it was previously disabled (as there are use cases for this.)

Comment (ksivaman) [reply transformer_engine/pytorch/module/layernorm_linear.py:412]: @sanandaraj5597 Any update here?

**Review by ksivaman — APPROVED** [review]
